dynamic-check/src/objects/frame.rs

```diff
@@ -0,0 +1,1198 @@
+
+        Ok(left_type)
+    }
+
+    ///
+    /// Returns the `Type` of the expression after the binary operation.
+    ///
+    /// Asserts that the `Type` is an integer.
+    ///
+    fn parse_integer_binary_expression(
+        &mut self,
+        left: &Expression,
+        right: &Expression,
+        span: &Span,
+    ) -> Result<Type, FrameError> {
+        let type_ = self.parse_binary_expression(left, right, span)?;
+
+        // Assert that the type is an integer.
+        self.assert_integer(&type_, span);
+
+        Ok(type_)
+    }
+
+    ///
+    /// Returns the `Boolean` type if the expression is a `Boolean` type.
+    ///
+    fn parse_boolean_expression(&mut self, expression: &Expression, span: &Span) -> Result<Type, FrameError> {
+        // Return the `Boolean` type
+        let boolean_type = Type::Boolean;
+
+        // Get the type of the expression
+        let expression_type = self.parse_expression(expression)?;
+
+        // Assert that the type is a boolean.
+        self.assert_equal(boolean_type.clone(), expression_type, span);
+
+        Ok(boolean_type)
+    }
+
+    ///
+    /// Returns the `Type` of the expression being negated. Must be a negative integer type.
+    ///
+    fn parse_negate_expression(&mut self, expression: &Expression, span: &Span) -> Result<Type, FrameError> {
+        // Parse the expression type.
+        let type_ = self.parse_expression(expression)?;
+
+        // Assert that this integer can be negated.
+        self.assert_negative_integer(&type_, span);
+
+        Ok(type_)
+    }
+
+    ///
+    /// Returns the `Boolean` type if the binary expression is a `Boolean` type.
+    ///
+    fn parse_boolean_binary_expression(
+        &mut self,
+        left: &Expression,
+        right: &Expression,
+        span: &Span,
+    ) -> Result<Type, FrameError> {
+        // Create the `Boolean` type.
+        let boolean_type = Type::Boolean;
+
+        // Create a new type assertion for the binary expression
+        let _binary_expression_type = self.parse_binary_expression(left, right, span)?;
+
+        // Return the `Boolean` type.
+        Ok(boolean_type)
+    }
+
+    ///
+    /// Returns the type of the conditional expression.
+    ///
+    fn parse_conditional_expression(
+        &mut self,
+        condition: &Expression,
+        first: &Expression,
+        second: &Expression,
+        span: &Span,
+    ) -> Result<Type, FrameError> {
+        // Check that the type of the condition expression is a boolean.
+        let _condition_type = self.parse_boolean_expression(condition, span)?;
+
+        // Check that the types of the first and second expression are equal.
+        self.parse_binary_expression(first, second, span)
+    }
+
+    ///
+    /// Returns the type of the tuple expression.
+    ///
+    fn parse_tuple(&mut self, expressions: &Vec<Expression>, _span: &Span) -> Result<Type, FrameError> {
+        let mut types = vec![];
```

Comment:
if the argument can be `ExactSizeIterator` (which is a more practical, but a bit more restrictive iterator), then this line could be
```suggestion
        let mut types = Vec::with_capacity(expressions.len());
```